Library/Homebrew/extend/os/mac/keg_relocate.rb

```diff
@@ -28,12 +28,14 @@ def relocate_dynamic_linkage(relocation)
 
         each_linkage_for(file, :dynamically_linked_libraries) do |old_name|
           new_name = relocated_name_for(old_name, relocation)
-          change_install_name(old_name, new_name, file) if new_name
+          new_name = loader_name_for(file, new_name)
+          change_install_name(old_name, new_name, file) if new_name != old_name
         end
 
         each_linkage_for(file, :rpaths) do |old_name|
           new_name = relocated_name_for(old_name, relocation)
-          change_rpath(old_name, new_name, file) if new_name
+          new_name = loader_name_for(file, new_name)
+          change_rpath(old_name, new_name, file) if new_name != old_name
```

Comment:
You're right. I had the wrong idea of what methods were called in each of the scenario. Looks like no changes to `#relocate_dynamic_linkage` are actually needed, for two reasons in fact:

1. `#fix_dynamic_linkage` is called anyway even when pouring from bottles (I didn't realize that). This makes it so that even "old" bottles get the relative names on install.

2. Bottles built after this change already have relative `install_name`s and `rpath`s, making much of the work done by `#relocate_dynamic_linkage` eventually unnecessary, as you mention.

**I've removed all changes from `#relocate_dynamic_linkage`, keeping just the changes to `#fix_dynamic_linkage`.**

> In fact, the rewriting absolute references to `$HOMEBREW_PREFIX` should make a significant chunk of `#relocate_dynamic_linkage` unnecessary. What do you think?

Indeed, after every bottle is rebuilt with this change, these lines should no longer be needed:

https://github.com/Homebrew/brew/blob/c5cbe6406b8b9d040c5d1850b5f041af4ddef83c/Library/Homebrew/extend/os/mac/keg_relocate.rb#L29-L37

(Some kind of deprecation cycle might be a good idea to let maintainers of third-party taps know that they should rebuild their bottles too)

Reply:
BTW, be aware that this PR assumes that `HOMEBREW_CELLAR` is a subdirectory of `HOMEBREW_PREFIX` (the actual name shouldn't matter). 

I mention it because I see that the relocation code deals `HOMEBREW_CELLAR` and `HOMEBREW_PREFIX` somewhat separately. I haven't seen a Homebrew install where `HOMEBREW_CELLAR != HOMEBREW_PREFIX/Cellar`, but if such an install exists, it will be broken by this PR.